src/BlockPatterns.php

```diff
@@ -0,0 +1,199 @@
+<?php
+namespace Automattic\WooCommerce\Blocks;
+
+use Automattic\WooCommerce\Blocks\Domain\Package;
+
+/**
+ * Registers patterns under the `./patterns/` directory.
+ * Each pattern is defined as a PHP file and defines its metadata using plugin-style headers.
+ * The minimum required definition is:
+ *
+ *     /**
+ *      * Title: My Pattern
+ *      * Slug: my-theme/my-pattern
+ *      *
+ *
+ * The output of the PHP source corresponds to the content of the pattern, e.g.:
+ *
+ *     <main><p><?php echo "Hello"; ?></p></main>
+ *
+ * Other settable fields include:
+ *
+ *   - Description
+ *   - Viewport Width
+ *   - Categories       (comma-separated values)
+ *   - Keywords         (comma-separated values)
+ *   - Block Types      (comma-separated values)
+ *   - Inserter         (yes/no)
+ *
+ * @internal
+ */
+class BlockPatterns {
+	const SLUG_REGEX            = '/^[A-z0-9\/_-]+$/';
+	const COMMA_SEPARATED_REGEX = '/[\s,]+/';
+
+	/**
+	 * Path to the patterns directory.
+	 *
+	 * @var string $patterns_path
+	 */
+	private $patterns_path;
+
+	/**
+	 * Constructor for class
+	 *
+	 * @param Package $package An instance of Package.
+	 */
+	public function __construct( Package $package ) {
+		$this->patterns_path = $package->get_path( 'patterns' );
+
+		add_action( 'init', array( $this, 'register_block_patterns' ) );
+	}
+
+	/**
+	 * Registers the block patterns and categories under `./patterns/`.
+	 */
+	public function register_block_patterns() {
+		if ( ! class_exists( 'WP_Block_Patterns_Registry' ) ) {
+			return;
+		}
+
+		$default_headers = array(
+			'title'         => 'Title',
+			'slug'          => 'Slug',
+			'description'   => 'Description',
+			'viewportWidth' => 'Viewport Width',
+			'categories'    => 'Categories',
+			'keywords'      => 'Keywords',
+			'blockTypes'    => 'Block Types',
+			'inserter'      => 'Inserter',
+		);
+
+		if ( file_exists( $this->patterns_path ) ) {
```

Comment:
[Optional] I'm sorry for being nitpicky here 😅 but I think we can also return early for this condition. I wasn't very clear in my previous comment. The reason that I care so much about indentation is the `register_block_patterns` is quite complex, so saving as many indentations as possible makes the code easier to read for me.